pkg/api/types.go

```diff
@@ -2119,34 +2119,121 @@ const (
 	StrategicMergePatchType PatchType = "application/strategic-merge-patch+json"
 )
 
-// Type and constants for component health validation.
+// Type and constants for component type.
+type ComponentType string
+
+// ComponentType constants define some known types of components.
+// Each type may have multiple instances, as long as their names are unique.
+const (
+	ComponentAPIServer         ComponentType = "apiserver"
+	ComponentControllerManager ComponentType = "controller-manager"
+	ComponentScheduler         ComponentType = "scheduler" // TODO: move mesos-specific component type & validation to contrib
+)
+
+// ComponentSpec describes the attributes that a component is created with.
+type ComponentSpec struct {
+	// Type of the component
+	Type ComponentType `json:"type"`
+	// Address to reach the component (scheme, host, port, path). Address paths determined by component type.
+	Address string `json:"address,omitempty"`
```

Comment:
Having a Probe type seems reasonable, with the following notes:
- It makes sense to support different probe protocols (tcp/http/https/ssl), ports, and paths (e.g. [AWS ELB](http://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/elb-healthchecks.html))
- There's no way to remotely probe via command execution (an existing probe type used by pods). So we'd likely have to deal with some naming overlap or figure out the commonalities between local and remote probes.
- It might make sense to have the Probe type on Service instead of on Component, to enable construction of advanced LoadBalencers.

Having a Service for each Component type seems reasonable, with the following notes:
- I wasn't sure if someone needed the root URL of the component (and not just the probe address). But if there's a Service for each component type then that might suffice.
- I wanted to support non-root paths for component addresses (e.g. in case the component is only accessible through a proxy that uses a path prefix). But perhaps this should be a Service enhancement instead.
- A Service (even with a LoadBalencer) probably still couldn't satisfy the desire to know if a Service or Component set was aggregately healthy/ready. Something still needs to produce a single boolean status from the set of component instances health/ready statuses based on some logic specific to that component (e.g. X is healthy if at least N instances are healthy).


In response to...
> A Service would need to be created for each component, but we plan to do that in the normal case, anyway.

I haven't heard this plan. The apiserver makes sense, but what other reasons are there to have services for each component (or one for each component type)?

In response to...
> The combination of the Component and Endpoints would provide presence monitoring.
>
> Self-registration would be a matter of updating the Endpoints resource corresponding to the component. Self-hosted components would not need to self-register.

I'm not sure I follow. Can you explain?